httpql/src/main/java/com/hubspot/httpql/internal/CombinedFilterEntry.java

```diff
@@ -0,0 +1,107 @@
+package com.hubspot.httpql.internal;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.jooq.Condition;
+import org.jooq.Operator;
+import org.jooq.impl.DSL;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.hubspot.httpql.FieldFactory;
+import com.hubspot.httpql.QuerySpec;
+
+public class CombinedFilterEntry<T extends QuerySpec> implements FilterEntryConditionCreator<T> {
+
+  private Operator operator;
+
+  private List<FilterEntryConditionCreator<T>> conditionCreators;
+
+  public CombinedFilterEntry(Operator operator, List<FilterEntryConditionCreator<T>> conditionCreators) {
+    this.operator = operator;
+    this.conditionCreators = Lists.newArrayList(conditionCreators);
+  }
+
+  @Override
+  public Condition getCondition(QuerySpec value, FieldFactory fieldFactory) {
+    return DSL.condition(operator, conditionCreators.stream()
+        .map(cc -> cc.getCondition(value, fieldFactory))
+        .collect(Collectors.toList()));
+  }
+
+  @Override
+  public Collection<BoundFilterEntry<T>> getFlattenedBoundFilterEntries() {
+    return conditionCreators.stream()
+        .map(FilterEntryConditionCreator::getFlattenedBoundFilterEntries)
+        .flatMap(Collection::stream)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  public void addConditionCreator(FilterEntryConditionCreator<T> conditionCreator) {
+    if (!getFlattenedBoundFilterEntries().contains(conditionCreator)) {
+      conditionCreators.add(conditionCreator);
+    }
+  }
+
+  public boolean removeAllFiltersFor(String fieldName) {
+    boolean removedFromThis = conditionCreators.removeIf(f -> f instanceof BoundFilterEntry && ((BoundFilterEntry<T>) f)
+        .getQueryName()
+        .equals(fieldName));
+    boolean removedFromChildren = conditionCreators.stream()
+        .filter(cc -> cc instanceof CombinedFilterEntry)
```

Comment:
It looks like you're doing this `instanceof` logic a lot to identify leaf nodes vs. combinations... would it be cleaner to add an `isCombinedCondition` helper (and/or `isLeafCondition`, or something like that) to `FilterEntryConditionCreator` that does the `instanceof` checks?

Reply:
I've removed most of these. I'll think about this some more but my knee jerk is that it seems odd to have helpers defined in the interface / higher level classes for checking if the instances are of a specific subclass type.

Reply:
Ahh sorry, I missed your follow up here. Yea that doesn't seem so bad. I guess I'm not sure what it would buy though in the implementation. Every instanceOf check I do is right before a cast and I wouldn't feel comfortable casting without that anyways.